### ISSUE TITLE
rule to ensure that default export matches module name

### ISSUE THREAD

**Issue by graingert** [benmosher/eslint-plugin-import#662]
It would be great to have a rule that  ensure that the default export matches module name:

eg

```js
// good
//doSomething.js

export default function doSomething {
    ...
}
```

```js
// bad
//dosomething.js

export default function doSomething {
    ...
}
```

**Comment by jfmengels** [comment]
I like the idea, though it's a really nitpicky style issue, as the name of the default import is rarely used anywhere else :sweat_smile:.

I usually name my files [in kebab-case](https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/filename-case.md), as I like it that way, and because apparently that causes less problems/confusion for people with case-insensitive operating systems (--> mac users).

I think we could simply transform the name of the file [to camelCase](https://lodash.com/docs/4.16.6#camelCase) and see if that matches the exported value. In the case files with dots in their name (`foo-bar.helper.js`), we could check if the name matches any part of the transformed name (`fooBar`, `helper` or `fooBarHelper`). For `index.js` files, I'd recommend to skip the check though.

If the exported value has no name (=== arrow function, object/array/primitive/...), then I assume this is fin to ignore?

**Comment by graingert** [comment]
> I like the idea, though it's a really nitpicky style issue, as the name of the default import is rarely used anywhere else :sweat_smile:.

I like to use:
```
import doSomething from './doSomething';
```

```
//doSomething.js
export default function doSomething() {
   ...
}
```

everywhere

**Comment by jfmengels** [comment]
True, but that's more of an import style issue (partially export too of course). You may be interested in this discussion https://github.com/sindresorhus/eslint-plugin-unicorn/issues/6.

**Comment by graingert** [comment]
@jfmengels I've seen https://github.com/sindresorhus/eslint-plugin-unicorn/blob/master/docs/rules/filename-case.md but it doesn't work for me, because I want:


```
// FooBar.js

export default class FooBar {
}
```

to work too.

**Comment by graingert** [comment]
@jfmengels sorry I posted my earlier comment before I'd finished it. See the completed version now...

**Comment by ljharb** [comment]
Also, `foo/index.js` should have an import named "foo" (following the same casing rules as if it was `foo.js`)

**Comment by graingert** [comment]
@ljharb `foo/index.js` can have a default export named "foo", if the default export has a name.

**Comment by ljharb** [comment]
oops yes sorry, default export, not import. totally agreed.

**Comment by benmosher** [comment]
I think it will be extraordinarily difficult to come to any consensus on a versatile implementation.

That said: I wouldn't use this rule. I'm still not super into the style guide enforcement aspect of import linting. So feel free to ignore my curmudgeonry, I'm not the target user of this rule. 😎

If y'all are able to figure out something you think would be both useful and not cause a million nitpick issues to be filed, I'm cool with it. I'm concerned that it will be something pretty niche that people will want to customize, which makes me lean towards this being is a spin-off module.

**Comment by graingert** [comment]
I'm pretty much in agreement with @ljharb

I'm also OK with a config that chooses between an exact match or camel-case, or etc

**Comment by peet** [comment]
I believe that [eslint-plugin-filenames](https://www.npmjs.com/package/eslint-plugin-filenames) already does what is being suggested here

**Comment by graingert** [comment]
@peet that looks about right

**Comment by graingert** [comment]
@ljharb https://github.com/airbnb/javascript/issues/1185

**Comment by wh1t3cAt1k** [comment]
I just wanted to let you guys know that eslint-plugin-filenames has been abandoned according to the author. I am one of those guys who would like to ensure that the exported entity name (even for non-default exports) matches the filename.

**Comment by minseoksuh** [comment]
I created a plugin "eslint-plugin-consistent-default-export-name" ([https://github.com/minseoksuh/eslint-plugin-consistent-default-export-name](https://github.com/minseoksuh/eslint-plugin-consistent-default-export-name))
that  combines  "match-exported" rule from `eslint-plugin-filenames` and "default-import-match-filename" rule from this repo. 

Try it out if you want to.

**Comment by ljharb** [comment]
@minseoksuh itd be great if you were willing to make a PR to eslint-plugin-import.

**Comment by minseoksuh** [comment]
@ljharb thanks for the response.
Do you mean I make a PR to eslint-plugin-import for the rule "match-exported"? (which in my plugin is named `default-export-match-filename`)

If it is ok with the theme of the repo "eslint-plugin-import", than I would love to do it.
In fact, I think much more people will be able to use, if the rules are in a popular plugin.

Let me know.
Thanks!

**Comment by ljharb** [comment]
@minseoksuh yes, if you can handle all the edge cases and have good test coverage, it seems like a great rule to me :-)

**Comment by uncaught** [comment]
@minseoksuh I'd prefer your new names, they are more explicit 👍
